nengo_gui/static/test/datastore_test.js

```diff
@@ -0,0 +1,201 @@
+    var data_store = new Nengo.DataStore(2, sim, 0);
+
+    beforeEach(function() {
+        data_store = new Nengo.DataStore(2, sim, 0);
+    });
+
+    it("accepts data", function() {
+        data_store.push([0.0, 1.1, 1.2]);
+        assert.deepEqual(data_store.data, [[1.1], [1.2]]);
+    });
+
+    it("orders data", function() {
+        data_store.push([1.0, 3.1, 3.2]);
+        data_store.push([0.5, 2.1, 2.2]);
+        assert.deepEqual(data_store.data, [[2.1], [2.2]]);
+    });
+
+    it("filters data", function() {
+        data_store.synapse = 0.1;
+        data_store.push([0.1, 2.1, 2.2]);
+        data_store.push([0.2, 3.1, 3.2]);
+        data_store.push([0.3, 1.1, 1.2]);
+        for(var i = 0; i < data_store.data.length; i++){
+            for(var j = 0; j < data_store.data[i].length; j++){
+                data_store.data[i][j] = Math.ceil(data_store.data[i][j]);
+            }
+        }
+        assert.deepEqual(data_store.data, [[3, 3, 2], [3, 3, 2]]);
+    });
+
+    it("resets data", function() {
+        data_store.push([0.5, 2.1, 2.2]);
+        data_store.push([1.0, 3.1, 3.2]);
+        data_store.push([1.5, 4.1, 4.2]);
+        data_store.reset();
+        assert.deepEqual(data_store.data, [[], []]);
+    });
+
+    it("throws away old data", function() {
+        data_store.push([0.5, 2.1, 2.2]);
+        data_store.push([1.0, 3.1, 3.2]);
+        data_store.push([1.5, 4.1, 4.2]);
+        data_store.sim.time_slider.last_time = 3;
+        data_store.update();
+        assert.deepEqual(data_store.data, [[3.1, 4.1], [3.2, 4.2]]);
+    });
+
+    it("gives the shown data", function() {
+        data_store.push([0.5, 2.1, 2.2]);
+        data_store.push([1.0, 3.1, 3.2]);
+        data_store.push([1.5, 4.1, 4.2]);
+        data_store.push([2.5, 5.1, 5.2]);
+        data_store.sim.time_slider.first_shown_time = 1.0;
+        data_store.sim.time_slider.shown_time = 1.0;
+        assert.deepEqual(data_store.get_shown_data(), [[3.1, 4.1], [3.2, 4.2]]);
+    });
+
+    it("gives the last data", function() {
+        data_store.push([0.5, 2.1, 2.2]);
+        data_store.push([1.0, 3.1, 3.2]);
+        data_store.push([1.5, 4.1, 4.2]);
+        data_store.push([2.5, 5.1, 5.2]);
+        data_store.sim.time_slider.first_shown_time = 1.0;
+        data_store.sim.time_slider.shown_time = 1.0;
+        assert.deepEqual(data_store.get_last_data(), [5.1, 5.2]);
+    });
+
+});
+
+describe("VariableDataStore", function() {
+    var data_store = new Nengo.VariableDataStore(2, sim, 0);
+
+    beforeEach(function() {
+        data_store = new Nengo.VariableDataStore(2, sim, 0);
+    });
+
+    it("accepts data", function() {
+        data_store.push([0.0, 1.1, 1.2])
+        assert.deepEqual(data_store.data, [[1.1], [1.2]])
+    });
+
+    it("accepts jagged data", function() {
+        data_store.push([0.0, 1.1, 1.2])
+        data_store.dims = 3;
+        assert.equal(data_store.dims, 3)
+        assert.equal(data_store._old_dims, 2)
```

Comment:
Is the underscore in `_old_dims` supposed to indicate a protected attribute as in Python? In that case it should be considered an implementation detail and not be tested.

Why is it stored anyways? It seems to be only set, but never read except in these tests.